### PR TITLE
Complete implementation with 100% coverage

### PR DESCRIPTION
**Pull request by nitika-goel**
Summary of changes made in contract:
- changed balanceOf function to keep in line with stndard(locked tokens are ignored in balanceOf)
- changed transfer and transferFrom to standard implementations
- added return type to lock function to maintain standard
- changes in data mapping and variable naming

Other changes
- added test cases
- added travis

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,16 @@
+sudo: required
+dist: trusty
+language: node_js
+node_js:
+  - '7'
```
Comment (davesag) [inline .travis.yml]: `node 8.11.3` is the current Long Term Supported release so I recommend testing against that.

File: .travis.yml [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+install:
+  - npm install -g truffle
+  - npm install -g ganache-cli
```
Comment (davesag) [inline .travis.yml]: not sure you need this as truffle actually includes ganache.

File: contracts/LockableToken.sol [hunk 3]
```diff
@@ -2,19 +2,16 @@ pragma solidity ^0.4.24;
...
 import 'openzeppelin-solidity/contracts/math/SafeMath.sol';
+// import './SafeMath.sol';
 
 
 contract LockableToken {
     using SafeMath for uint256;
 
     /**
-     * @dev Utilities for which a token can be locked
+     * @dev Utilities for which a user's token have been locked
```
Comment (davesag) [inline contracts/LockableToken.sol]: Maybe change the wording to "Reasons why a user's tokens have been locked"

File: contracts/LockableToken.sol [hunk 4]
```diff
@@ -49,68 +46,118 @@ contract LockableToken {
...
-        locked_for.push("CA");
-        locking_active["CA"]=true;
-        //Utility  - Governance
-        locked_for.push("GOV");
-        locking_active["GOV"]=true;
+    }
+    
+    /**
+     * @dev Locks a specified amount of tokens against an address,
+     *      for a specified purpose and time
```
Comment (davesag) [inline contracts/LockableToken.sol]: change 'purpose' to 'reason' to be consistent

File: contracts/LockableToken.sol [hunk 5]
```diff
@@ -49,68 +46,118 @@ contract LockableToken {
...
+     * @param _reason The purpose to lock tokens
```
Comment (davesag) [inline contracts/LockableToken.sol]: change to "The reason the tokens are locked"

File: contracts/LockableToken.sol [hunk 6]
```diff
@@ -49,68 +46,118 @@ contract LockableToken {
...
+     * @param _amount Number of tokens to be locked
+     * @param _time Lock time in seconds
+     */
+    function lock(bytes32 _reason, uint256 _amount, uint256 _time)
+        public
+        returns (bool)
+    {
+        uint256 validUntil=block.timestamp.add(_time);
```
Comment (davesag) [inline contracts/LockableToken.sol]: space before and after `=`

File: contracts/LockableToken.sol [hunk 7]
```diff
@@ -49,68 +46,118 @@ contract LockableToken {
...
+        if (locked[msg.sender][_reason].amount == 0)
+            lockReason[msg.sender].push(_reason);
+        locked[msg.sender][_reason] = lockToken(_amount, validUntil);
+        emit Lock(msg.sender, _reason, _amount, validUntil);
+        return true;
     }
 
     /**
      * @dev Returns tokens locked for a specified address for a
      *      specified purpose at a specified time
```
Comment (davesag) [inline contracts/LockableToken.sol]: change 'purpose' to 'reason' to be consistent

File: test/LockableToken-creation_test.js [hunk 8]
```diff
@@ -1,14 +1,18 @@
...
 const assertThrows = require('./utils/assertThrows')
 const assertExpectedArguments = require('./utils/assertExpectedArguments')
+const { assertRevert } = require('./utils/assertRevert')
 
-contract('LockableToken', ([owner]) => {
+contract('LockableToken', ([owner, receiver, spender]) => {
   const supply = 1000
-  const lockedFor = 'CA'
-  const lockedAmount = 500
+  const lockReason =
```
Comment (davesag) [inline test/LockableToken-creation_test.js]: In JS these can actually just be strings and `web3` will handle the conversion to `bytes32`.  so you could say
```
const lockReason1 = 'some reason'
const lockReason2 = 'some other reason'
```

File: test/LockableToken-creation_test.js [hunk 9]
```diff
@@ -27,21 +31,153 @@ contract('LockableToken', ([owner]) => {
...
+      )
+      assert.equal(
+        increasedLockAmount.toNumber(),
+        actualLockedAmount.toNumber() + lockedAmount
+      )
+    })
+
+    it('cannot transfer tokens to null address', async function() {
+      await assertRevert(
+        token.transfer(0x0000000000000000000000000000000000000000, 100, {
```
Comment (davesag) [inline test/LockableToken-creation_test.js]: make `0x0000000000000000000000000000000000000000` a constant.

**Review by davesag — CHANGES_REQUESTED** [review]
Cool.  Just a few wording and stylistic changes requested.

**Review by davesag — APPROVED** [review]

**Comment by nitika-goel** [comment]
@davesag , thanks for reviewing. Have made a few changes in comments, updated readme and incorporated the changes highlighted. Let me know what you think.

**Pull request merged**